File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.js
function solution(diffs, times, limit) {
    let left = 1
    let right = 100000
    let mid = undefined
    let answer = right
    
    while (left <= right) {
        mid = Math.floor((left + right) / 2)
        let time = 0, flag = false
        for (let i = 0; i<diffs.length; ++i) {
            if (mid - diffs[i] < 0) {
                time = time + (diffs[i]-mid) * (times[i] + times[i-1]) + times[i]
            } else {
                time += times[i]
            }
            
            if (limit < time) {
                flag = true
                break
            }
        }
        
        if (flag) {
            left = mid + 1
        } else {
            answer = mid
            right = mid - 1
        }
    } 
        
    return answer;
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    answer = float("inf")
    
    left, right = 1, 1000001 
    
    while left <= right:
        mid = (left + right) // 2  
        total_time = 0 
        flag = True 
        
        for i in range(len(diffs)):
            if diffs[i] > mid:
                mistakes = diffs[i] - mid
                if i == 0:
                    total_time += times[i] * (mistakes + 1)
                else:
                    total_time += (times[i] + times[i-1]) * mistakes + times[i]
            else:
                total_time += times[i]
                
            if total_time > limit:
                flag = False
                break
            
        if flag:
            answer = mid
            right = mid - 1
        else:
            left = mid + 1
    
    return answer
